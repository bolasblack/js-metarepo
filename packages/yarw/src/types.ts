import { Root, RootState } from './Root'

/**
 * A normal [action](https://redux.js.org/basics/actions)
 */
export interface Action<T = unknown, P = undefined> {
  type: T
  payload: P
}
export namespace Action {
  export type Payload<P, T = unknown> = Action<T, P>
  export type Empty<T = unknown> = Action<T>
  export type Any<T = unknown> = Action<T, any>
}
export type ActionFromCaseReducer<
  R,
  ActionType extends string
> = R extends CaseReducer<any, infer A> ? A & { type: ActionType } : never
export type ActionAddPrefix<
  A extends Action<string>,
  ActionTypePrefix extends string
> = Action<`${ActionTypePrefix}/${TypeFromAction<A>}`, PayloadFromAction<A>>

export type ActionMap = {
  [type: string]: Action.Any<string>
}
export type ActionMapFromCaseReducerMap<Rs, ActionTypePrefix extends string> = {
  [type in keyof Rs]: type extends string
    ? ActionFromCaseReducer<Rs[type], `${ActionTypePrefix}/${type}`>
    : never
}
export type ActionMapAddPrefix<
  As extends ActionMap,
  ActionTypePrefix extends string
> = {
  [K in keyof As]: ActionAddPrefix<As[K], ActionTypePrefix>
}

export type TypeFromAction<A> = A extends Action<infer T> ? T : never
export type PayloadFromAction<A> = A extends Action<any, infer P> ? P : never

/**
 * A function to create an [[Action]]
 */
export type ActionCreator<A extends Action.Any> = A extends Action.Empty
  ? ActionCreator.ForEmptyAction<A>
  : ActionCreator.ForPayloadAction<A>
export namespace ActionCreator {
  /**
   * A [[ActionCreator]] to create an non-payload [[Action]]
   */
  export interface ForEmptyAction<A extends Action.Any> {
    (): A
  }

  /**
   * A [[ActionCreator]] to create an payload [[Action]]
   */
  export interface ForPayloadAction<A extends Action.Any> {
    (payload: PayloadFromAction<A>): A
  }

  export type Any<A extends Action.Any = Action.Any> =
    | ActionCreator.ForEmptyAction<A>
    | ActionCreator.ForPayloadAction<A>
}

// 类型参数 S 用来在有些场景下推断 ActionDispatcher 所属 Slice 的 State 类型
/**
 * A function that can dispatch the corresponding action.
 */
export type ActionDispatcher<S, A extends Action.Any> = A extends Action.Empty
  ? ActionDispatcher.ForEmptyAction<S, A>
  : ActionDispatcher.ForPayloadAction<S, A>
export namespace ActionDispatcher {
  /** @hidden */
  const isActionDispatcherSym = Symbol('isActionDispatcher')

  /**
   * Create an [[ActionDispatcher]]
   *
   * @param [[Action.type]]
   * @param
   */
  export function create<S, A extends Action<any>>(
    type: TypeFromAction<A>,
    parent: Root,
  ): ActionDispatcher.Any<S, A> {
    const actionCreator = (...args: any[]): Action<A> => {
      return { type, payload: args[0] }
    }

    const dispatcher = (...args: any): void => {
      parent.dispatch((actionCreator as any)(...args))
    }

    dispatcher.match = (a: Action.Any): a is A => a.type === type

    dispatcher.create = actionCreator

    dispatcher.toString = () => String(type)

    dispatcher[isActionDispatcherSym] = true

    return dispatcher as ActionDispatcher<S, A>
  }

  /**
   * Check if `fn` is an [[ActionDispatcher]]
   */
  export function isActionDispatcher<S, A extends Action.Any>(
    fn: (...args: any) => any,
  ): fn is ActionDispatcher<S, A> {
    return fn[isActionDispatcherSym]
  }

  /** @hidden */
  interface Base<A extends Action.Any> {
    /**
     * Check if an [[Action]] matches the current [[ActionDispatcher]]
     */
    match(a: Action.Any): a is A

    /**
     * Create a new [[Action]]
     */
    create: ActionCreator<A>

    /**
     * Transform [[ActionDispatcher]] to [[Action.type]]
     */
    toString(): string
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars-experimental
  export interface ForEmptyAction<S, A extends Action.Any> extends Base<A> {
    (): void
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars-experimental
  export interface ForPayloadAction<S, A extends Action.Any> extends Base<A> {
    (payload: PayloadFromAction<A>): void
  }

  export type Any<S = any, A extends Action.Any = Action.Any> =
    | ActionDispatcher.ForEmptyAction<S, A>
    | ActionDispatcher.ForPayloadAction<S, A>
}
export type ActionDispatcherMap<S, As extends ActionMap> = {
  [type in keyof As]: ActionDispatcher<S, As[type]>
}

/**
 * A [Reducer](https://redux.js.org/basics/reducers) that never needs to
 * determine whether the parameter `state` is `undefined`
 */
export interface NaiveReducer<S, A extends Action.Any> {
  (state: S, action: A): S
}

/**
 * A [[NaiveReducer]] that only needs to process the corresponding [[Action]]
 */
export type CaseReducer<S, A extends Action.Any> = NaiveReducer<S, A>
export namespace CaseReducer {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars-experimental
  export const id = <S>(s: S, action: Action.Empty): S => s
}
export type CaseReducerMap<S> = {
  [type: string]: CaseReducer<S, Action.Any>
}

export interface ReducerBuilder {
  when: ReducerBuilder.When
}
export namespace ReducerBuilder {
  export type ActionMatcher = (action: Action.Any) => boolean

  export type Reducer<A extends Action.Any = Action.Any> = (
    state: RootState,
    action: A,
  ) => RootState

  export interface When {
    <A extends Action.Any = Action.Any>(reducer: Reducer<A>): void
    <A extends Action.Any = Action.Any>(
      actionFilter: ActionMatcher,
      reducer: Reducer<A>,
    ): void
    <A extends Action.Any = Action.Any>(
      actionDispatcher: ActionDispatcher.Any<any, A>,
      reducer: Reducer<A>,
    ): void
  }
}

export interface ActionListener<S = any> {
  (listener: ActionListener.ListenHandler): () => void
  <A extends Action.Any = Action.Any>(
    actionFilter: ActionListener.ListenMatcher<A>,
    listener: ActionListener.ListenHandler<A>,
  ): () => void
  <A extends Action.Any = Action.Any>(
    actionDispatcher: ActionDispatcher.Any<S, A>,
    listener: ActionListener.ListenHandler<A>,
  ): () => void
}
export namespace ActionListener {
  /* @hidden */
  export type ListenMatcher<A extends Action = Action.Any> = (
    action: Action.Any,
  ) => action is A

  /* @hidden */
  export type ListenHandler<A extends Action = Action.Any> = (action: A) => any

  /* @hidden */
  export const dispatchArgs = (
    dispatcher:
      | ActionDispatcher.Any
      | ActionListener.ListenMatcher
      | ActionListener.ListenHandler,
    callback: ActionListener.ListenHandler | undefined,
  ): {
    matcher: ActionListener.ListenMatcher | null
    handler: ActionListener.ListenHandler
  } => {
    let matcher: ActionListener.ListenMatcher | null
    let handler: ActionListener.ListenHandler

    if (!callback) {
      matcher = null
      handler = dispatcher
    } else {
      matcher = (a): a is any => {
        if (ActionDispatcher.isActionDispatcher(dispatcher)) {
          return dispatcher.match(a)
        } else {
          return dispatcher(a)
        }
      }
      handler = callback!
    }

    return { matcher, handler }
  }
}
