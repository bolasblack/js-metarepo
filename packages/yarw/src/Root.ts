import {
  Store,
  StoreEnhancer,
  createStore,
  Reducer,
  Observable,
  Middleware,
  applyMiddleware,
  isAction,
} from 'redux'
import $$observable from 'symbol-observable'
import { Action, NaiveReducer, ActionDispatcher, ActionListener } from './types'
import { composeNaiveReducers, composeEnhancers } from './utils'

export const internalStateSym = Symbol('internalState')

export type RegisterReducer = (
  reducer: NaiveReducer<RootState, Action.Any>,
) => () => void

export interface RootState {
  readonly [internalStateSym]: any
}

export interface Root {
  getState: Store<RootState>['getState']
  registerReducer: RegisterReducer
  listen: ActionListener
  dispatch: Store<RootState>['dispatch']
  subscribe: Store<RootState>['subscribe']
  [Symbol.observable](): Observable<RootState>
}

const defaultInitialState = Object.preventExtensions(
  Object.freeze({ [internalStateSym]: {} }),
)

export function createRoot(enhancer?: StoreEnhancer): Root {
  const { enhancer: listenEnhancer, listen } = createListenManager()
  const defaultReducer: Reducer<RootState> = s => s || defaultInitialState

  const store = createStore(
    defaultReducer,
    composeEnhancers(enhancer, listenEnhancer),
  )

  const { registerReducer } = createReducerManager(store)

  return {
    registerReducer,
    listen,
    dispatch: store.dispatch as any,
    subscribe: store.subscribe,
    getState: store.getState,
    [Symbol.observable]: (store as any)[$$observable],
    [$$observable as any]: (store as any)[$$observable],
  }
}

function createListenManager(): {
  enhancer: StoreEnhancer
  listen: ActionListener
} {
  let registered: {
    matcher: ActionListener.ListenMatcher
    handler: ActionListener.ListenHandler
  }[] = []

  const listen: ActionListener = function (
    dispatcher:
      | ActionDispatcher.Any
      | ActionListener.ListenMatcher
      | ActionListener.ListenHandler,
    callback?: ActionListener.ListenHandler,
  ) {
    const { matcher, handler } = ActionListener.dispatchArgs(
      dispatcher,
      callback,
    )
    const pair = { matcher: matcher || ((_a): _a is any => true), handler }
    registered.push(pair)
    return () => {
      registered = registered.filter(i => i !== pair)
    }
  }

  const middleware: Middleware = () => nextDispatch => action => {
    const result = nextDispatch(action)
    registered.forEach(i => {
      if (isAction(action) && i.matcher(action as any)) {
        try {
          i.handler(action as any)
        } catch (err) {
          console.error(err)
        }
      }
    })
    return result
  }

  return {
    listen,
    enhancer: applyMiddleware(middleware),
  }
}

function createReducerManager(store: Store<RootState>): {
  registerReducer: RegisterReducer
} {
  let registeredReducers: NaiveReducer<RootState, Action.Any>[] = []

  const registerReducer: RegisterReducer = r => {
    if (registeredReducers.includes(r)) {
      // TODO
      throw new Error()
    }

    updateReducers(registeredReducers.concat(r))

    return () => {
      updateReducers(registeredReducers.filter(_r => _r !== r))
    }
  }

  return { registerReducer }

  function createRootReducer(): Reducer<RootState, Action.Any> {
    const composedReducer = composeNaiveReducers(...registeredReducers)

    return (s, action) => {
      return composedReducer(s || defaultInitialState, action)
    }
  }

  function updateReducers(reducers: typeof registeredReducers): void {
    registeredReducers = reducers
    store.replaceReducer(createRootReducer() as any)
  }
}
