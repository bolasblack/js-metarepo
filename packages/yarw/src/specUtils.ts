// 帮助做一些简单的类型检查
// eslint-disable-next-line @typescript-eslint/no-empty-function
export function expectType<T>(_: T): void {}
