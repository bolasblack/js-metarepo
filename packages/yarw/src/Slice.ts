import { RootState, Root, internalStateSym } from './Root'
import {
  Action,
  ActionMap,
  ActionDispatcherMap,
  CaseReducerMap,
  ActionDispatcher,
  NaiveReducer,
  ActionListener,
  ReducerBuilder,
  ActionMapFromCaseReducerMap,
} from './types'
import { composeNaiveReducers } from './utils'

type StateFromCaseReducerMap<CRM> = CRM extends CaseReducerMap<infer S>
  ? S
  : never

type StateFromActionDispatcherMap<ADM> = ADM extends ActionDispatcherMap<
  infer S,
  any
>
  ? S
  : never

export interface Slice<S, As extends ActionMap> {
  actions: ActionDispatcherMap<S, As>
  reducer: NaiveReducer<RootState, Action.Any>
  listen: ActionListener<S>
  subscribe: Root['subscribe']
  getState: Slice.GetState<S>
  setState: Slice.SetState<S>
  unregister: () => void
  unregistered: boolean
}
export namespace Slice {
  export type FromSpec<
    Name extends string,
    Spec extends SliceSpec<any, any>,
  > = Slice<
    Spec['initialState'],
    ActionMapFromCaseReducerMap<Spec['reducers'], Name>
  >

  export type StateType<S extends Slice<any, any>> = S extends Slice<
    infer R,
    any
  >
    ? R
    : never

  export interface GetState<S> {
    (rootState?: RootState): S
  }

  export interface SetState<S> {
    (newState: S, rootState: RootState): RootState
    (newState: S): (rootState: RootState) => RootState
    (setter: (state: S) => S, rootState: RootState): RootState
    (setter: (state: S) => S): (rootState: RootState) => RootState
  }
}

export interface SliceSpec<S, Rs extends CaseReducerMap<S>> {
  initialState: S
  reducers: Rs
}

export function sliceSpec<S, Reducers extends CaseReducerMap<S>>(
  spec: SliceSpec<S, Reducers>,
): SliceSpec<S, Reducers> {
  return spec
}

export function registerSlice<
  N extends string,
  S,
  Reducers extends CaseReducerMap<S>,
>(
  parent: Root,
  name: N,
  spec: SliceSpec<S, Reducers>,
  extra?: (
    builder: ReducerBuilder,
    slice: Slice<S, ActionMapFromCaseReducerMap<Reducers, N>>,
  ) => void,
): Slice<S, ActionMapFromCaseReducerMap<Reducers, N>> {
  const unregisteredCtrl = new AbortController()

  // eslint-disable-next-line prefer-const
  let unregisterReducer: undefined | (() => void)

  type AM = ActionMapFromCaseReducerMap<Reducers, N>
  const actions: ActionDispatcherMap<S, AM> = createActionDispatchers(
    name,
    spec.reducers,
    parent,
    unregisteredCtrl.signal,
  )

  const listen: ActionListener<S> = createSliceListener(
    name,
    actions,
    parent,
    unregisteredCtrl.signal,
  )

  const { reducer, builder } = createSliceReducer(
    name,
    spec,
    actions,
    unregisteredCtrl.signal,
  )

  const wrappedGetState = (parentState?: RootState): S =>
    getState(
      name,
      spec,
      parentState ?? parent.getState(),
      unregisteredCtrl.signal,
    )

  const wrappedSetState = (
    newState: S | ((state: S) => S),
    parentState?: RootState,
  ): ((parentState: RootState) => RootState) | RootState => {
    if (!parentState) {
      return parentState => {
        return wrappedSetState(newState, parentState) as RootState
      }
    }

    if (typeof newState !== 'function') {
      return setState(name, newState, parentState, unregisteredCtrl.signal)
    } else {
      return setState(
        name,
        (newState as any)(wrappedGetState(parentState)),
        parentState,
        unregisteredCtrl.signal,
      )
    }
  }

  const slice: Slice<S, ActionMapFromCaseReducerMap<Reducers, N>> = {
    listen,
    actions,
    reducer,
    subscribe: (listener: () => void) => {
      if (unregisteredCtrl.signal.aborted) {
        console.error(
          `Calling unregistered slice "${name}"'s action dispatcher`,
        )
      }

      return parent.subscribe(listener)
    },
    getState: wrappedGetState,
    setState: wrappedSetState as Slice.SetState<S>,
    unregister: () => {
      if (unregisteredCtrl.signal.aborted) return
      unregisterReducer?.()
      unregisteredCtrl.abort()
    },
    get unregistered() {
      return unregisteredCtrl.signal.aborted
    },
  }

  extra?.(builder, slice)

  unregisterReducer = parent.registerReducer(reducer)

  return slice
}

const getState = <S>(
  sliceName: string,
  spec: SliceSpec<S, any>,
  rs: RootState,
  unregisterSignal: AbortSignal,
): S => {
  if (unregisterSignal.aborted) {
    console.error(
      `Calling unregistered slice "${sliceName}"'s action dispatcher`,
    )
  }

  return rs[internalStateSym][sliceName] || spec.initialState
}

const setState = <S>(
  sliceName: string,
  newState: S,
  rs: RootState,
  unregisterSignal: AbortSignal,
): RootState => {
  if (unregisterSignal.aborted) {
    console.error(
      `Calling unregistered slice "${sliceName}"'s action dispatcher`,
    )
  }

  return {
    ...rs,
    [internalStateSym]: {
      ...rs[internalStateSym],
      [sliceName]: newState,
    },
  }
}

function createSliceReducer<S>(
  sliceName: string,
  spec: SliceSpec<S, CaseReducerMap<S>>,
  actionDispatchers: ActionDispatcherMap<
    S,
    ActionMapFromCaseReducerMap<CaseReducerMap<S>, typeof sliceName>
  >,
  unregisterSignal: AbortSignal,
): {
  builder: ReducerBuilder
  reducer: NaiveReducer<RootState, Action.Any>
} {
  const extraReducers: NaiveReducer<RootState, Action.Any>[] = []

  return {
    builder: { when },
    reducer: (rs, a) =>
      composeNaiveReducers(
        createReducerFromSpec(
          sliceName,
          spec,
          actionDispatchers,
          unregisterSignal,
        ),
        ...extraReducers,
      )(rs, a),
  }

  function when(
    matcher:
      | ActionDispatcher.Any
      | ReducerBuilder.ActionMatcher
      | ReducerBuilder.Reducer,
    reducer?: ReducerBuilder.Reducer,
  ): void {
    let _matcher: ReducerBuilder.ActionMatcher
    let _reducer: ReducerBuilder.Reducer

    if (!reducer) {
      _matcher = () => true
      _reducer = matcher as ReducerBuilder.Reducer
    } else if (ActionDispatcher.isActionDispatcher(matcher)) {
      _matcher = matcher.match
      _reducer = reducer
    } else {
      _matcher = matcher as ReducerBuilder.ActionMatcher
      _reducer = reducer
    }

    extraReducers.push((rs, a) => (_matcher(a) ? _reducer(rs, a) : rs))
  }
}

function createReducerFromSpec<S>(
  sliceName: string,
  spec: SliceSpec<S, CaseReducerMap<S>>,
  actionDispatchers: ActionDispatcherMap<
    S,
    ActionMapFromCaseReducerMap<CaseReducerMap<S>, typeof sliceName>
  >,
  unregisterSignal: AbortSignal,
): NaiveReducer<RootState, Action.Any> {
  const matchers: {
    name: string
    match: (a: Action.Any) => boolean
  }[] = Object.keys(actionDispatchers).map(name => ({
    name,
    match: actionDispatchers[name].match,
  }))

  return (s, a) => {
    if (unregisterSignal.aborted) {
      console.error(
        `Calling unregistered slice "${sliceName}"'s action dispatcher`,
      )
    }

    const caseReducers = matchers
      .filter(m => m.match(a))
      .map(m => spec.reducers[m.name])

    if (!caseReducers.length) return s

    const reducer = composeNaiveReducers(...caseReducers)
    const oldSelfState = getState(sliceName, spec, s, unregisterSignal)
    const newSelfState = reducer(oldSelfState, a)
    return setState(sliceName, newSelfState, s, unregisterSignal)
  }
}

function createSliceListener<As extends ActionDispatcherMap<any, any>>(
  sliceName: string,
  actions: As,
  parent: Root,
  unregisterSignal: AbortSignal,
): ActionListener<StateFromActionDispatcherMap<As>> {
  const actionMatchers = Object.values(actions).map(ad => ad.match)
  const defaultMatcher = (a: Action.Any): a is Action.Any =>
    actionMatchers.some(m => m(a))

  return function (
    dispatcher:
      | ActionDispatcher.Any
      | ActionListener.ListenMatcher
      | ActionListener.ListenHandler,
    callback?: ActionListener.ListenHandler,
  ) {
    if (unregisterSignal.aborted) {
      console.error(
        `Calling unregistered slice "${sliceName}"'s action dispatcher`,
      )
    }

    const { matcher, handler } = ActionListener.dispatchArgs(
      dispatcher,
      callback,
    )

    return parent.listen(matcher || defaultMatcher, handler)
  }
}

function createActionDispatchers<
  N extends string,
  Reducers extends CaseReducerMap<any>,
>(
  sliceName: N,
  reducers: Reducers,
  parent: Root,
  unregisterSignal: AbortSignal,
): ActionDispatcherMap<
  StateFromCaseReducerMap<Reducers>,
  ActionMapFromCaseReducerMap<Reducers, N>
> {
  type Ret = ActionDispatcherMap<
    StateFromCaseReducerMap<Reducers>,
    ActionMapFromCaseReducerMap<Reducers, N>
  >

  const actionTypes = Object.keys(reducers) as StringOnly<keyof Reducers>[]

  return actionTypes.reduce((actions, actionType) => {
    const dispatcher: any = ActionDispatcher.create(
      `${sliceName}/${actionType}`,
      parent,
    )

    actions[actionType] = ((...args: any) => {
      if (unregisterSignal.aborted) {
        console.error(
          `Calling unregistered slice "${sliceName}"'s action dispatcher`,
        )
      }

      return dispatcher(...args)
    }) as any
    Object.assign(actions[actionType], dispatcher)

    return actions
  }, {} as Ret)
}

type StringOnly<T> = T extends string ? T : never
