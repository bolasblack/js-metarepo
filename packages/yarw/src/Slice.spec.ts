import 'console-testing-library'
import { sliceSpec, Slice, registerSlice } from './Slice'
import { createRoot, Root, internalStateSym } from './Root'
import { Action, CaseReducer } from './types'
import { expectType } from './specUtils'

describe('sliceSpec', () => {
  it('return spec directly', () => {
    const fakeSpec = Symbol('fakeSpec')
    expect(sliceSpec(fakeSpec as any)).toBe(fakeSpec)
  })

  it('check and infer spec type', () => {
    type State = { test: number }

    const spec = sliceSpec({
      initialState: { test: 1 },
      reducers: {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars-experimental
        a(s, a: Action.Empty) {
          expectType<State>(s)
          return s
        },
        // eslint-disable-next-line @typescript-eslint/no-unused-vars-experimental
        b(s, a: Action.Payload<number>) {
          expectType<State>(s)
          return s
        },
      },
    })

    expectType<State>(spec.initialState)
    expectType<{
      a: (s: State, a: Action.Empty) => State
      b: (s: State, a: Action.Payload<number>) => State
    }>(spec.reducers)
  })
})

describe('registerSlice', () => {
  let root: Root

  beforeEach(() => {
    root = createRoot()
  })

  it('create a Slice', () => {
    const slice = registerSlice(root, 'test', {
      initialState: Symbol('initial'),
      reducers: {
        set(s, a: Action.Payload<symbol>) {
          return a.payload
        },
      },
    })

    expect(slice).toEqual(
      expect.objectContaining({
        actions: { set: expect.any(Function) },
        reducer: expect.any(Function),
        listen: expect.any(Function),
        subscribe: expect.any(Function),
        getState: expect.any(Function),
        unregister: expect.any(Function),
        unregistered: expect.any(Boolean),
      }),
    )
  })

  describe('argument extra', () => {
    it('works', () => {
      let extraArgSlice: Slice<any, any> | null = null

      const slice1InitialState = Symbol('slice1 init')
      const slice1UpdatedState = Symbol('slice1 updated')
      const slice1 = registerSlice(root, '1', {
        initialState: slice1InitialState,
        reducers: {
          set(s, a: Action.Payload<symbol>) {
            return a.payload
          },
        },
      })

      const slice2 = registerSlice(
        root,
        '2',
        {
          initialState: {
            prevSlice1State: slice1.getState(),
            prevSlice1State$set: null as any,
          },
          reducers: { init: CaseReducer.id },
        },
        (builder, slice) => {
          extraArgSlice = slice

          builder.when((rs) =>
            slice.setState(
              {
                ...slice.getState(rs),
                prevSlice1State: slice1.getState(rs),
              },
              rs,
            ),
          )

          builder.when(slice1.actions.set, (rs, a) =>
            slice.setState(
              {
                ...slice.getState(rs),
                prevSlice1State$set: a.payload,
              },
              rs,
            ),
          )
        },
      )

      expect(extraArgSlice).toBe(slice2)
      expect(slice2.getState()).toEqual({
        prevSlice1State: slice1InitialState,
        prevSlice1State$set: null,
      })

      slice1.actions.set(slice1UpdatedState)
      expect(slice2.getState()).toEqual({
        prevSlice1State: slice1InitialState,
        prevSlice1State$set: slice1UpdatedState,
      })
    })
  })
})

describe('Slice', () => {
  let root: Root
  let slice: Slice<symbol, { set: Action.Payload<symbol> }>
  const sliceName = 'test' as const
  const initialState = Symbol('initial')
  const updatedState = Symbol('updated')

  beforeEach(() => {
    root = createRoot()

    slice = registerSlice(root, sliceName, {
      initialState,
      reducers: {
        set(s, a: Action.Payload<symbol>) {
          return a.payload
        },
      },
    })
  })

  describe('actions', () => {
    it('can generate right type', () => {
      const root = createRoot()

      const slice = registerSlice(root, sliceName, {
        initialState,
        reducers: {
          set(s, a: Action.Payload<symbol>) {
            return a.payload
          },
        },
      })

      expectType<(payload: symbol) => void>(slice.actions.set)
      expectType<void>(slice.actions.set(updatedState))
      expectType<{ type: 'test/set'; payload: symbol }>(
        slice.actions.set.create(updatedState),
      )
    })

    it('can dispatch self action', () => {
      root.dispatch = jest.fn(root.dispatch)

      slice.actions.set(updatedState)

      expect(root.dispatch).toBeCalledWith({
        type: `${sliceName}/set`,
        payload: updatedState,
      })

      expect(root.getState()).toEqual({
        [internalStateSym]: {
          [sliceName]: updatedState,
        },
      })
    })
  })

  describe('reducer', () => {
    it('works', () => {
      expect(root.getState()).toEqual({ [internalStateSym]: {} })

      const anotherSym = Symbol('updated2')
      const newState = slice.reducer(root.getState(), {
        type: `${sliceName}/set`,
        payload: anotherSym,
      })
      expect(newState).toEqual({
        [internalStateSym]: { [sliceName]: anotherSym },
      })

      expect(root.getState()).toEqual({ [internalStateSym]: {} })
    })
  })

  describe('listen', () => {
    it('listen self actions by default', () => {
      const listener = jest.fn()
      const helloAction = { type: 'hello' }
      slice.listen(listener)

      root.dispatch(helloAction)
      expect(listener).not.toBeCalledWith(helloAction)

      slice.actions.set(updatedState)
      expect(listener).toBeCalledWith({
        type: slice.actions.set.toString(),
        payload: updatedState,
      })
    })
  })

  describe('subscribe', () => {
    it('callback when any action dispatched in parent', () => {
      const listener = jest.fn()
      const helloAction = { type: 'hello' }
      slice.subscribe(listener)

      root.dispatch(helloAction)
      expect(listener).lastCalledWith()
      expect(listener).toBeCalledTimes(1)

      slice.actions.set(updatedState)
      expect(listener).lastCalledWith()
      expect(listener).toBeCalledTimes(2)
    })
  })

  describe('getState', () => {
    it('works', () => {
      expect(root.getState()).toEqual({ [internalStateSym]: {} })
      expect(slice.getState()).toEqual(initialState)
      slice.actions.set(updatedState)
      expect(root.getState()).toEqual({
        [internalStateSym]: { [sliceName]: updatedState },
      })
      expect(slice.getState()).toEqual(updatedState)

      const anotherSym = Symbol('fake state')
      expect(slice.getState({ [internalStateSym]: {} })).toEqual(initialState)
      expect(
        slice.getState({ [internalStateSym]: { [sliceName]: anotherSym } }),
      ).toEqual(anotherSym)
    })
  })

  describe('setState', () => {
    it('works', () => {
      expect(slice.setState(updatedState, { [internalStateSym]: {} })).toEqual({
        [internalStateSym]: { [sliceName]: updatedState },
      })
    })

    it('receive a setter function as first argument', () => {
      expect(
        slice.setState(() => updatedState, { [internalStateSym]: {} }),
      ).toEqual({
        [internalStateSym]: { [sliceName]: updatedState },
      })
    })

    it('is a curried function', () => {
      expect(slice.setState(updatedState)).toEqual(expect.any(Function))
      expect(slice.setState(updatedState)({ [internalStateSym]: {} })).toEqual({
        [internalStateSym]: { [sliceName]: updatedState },
      })

      expect(slice.setState(() => updatedState)).toEqual(expect.any(Function))
      expect(
        slice.setState(() => updatedState)({ [internalStateSym]: {} }),
      ).toEqual({
        [internalStateSym]: { [sliceName]: updatedState },
      })
    })
  })

  describe('unregister', () => {
    it('should console error when calling `getState` after unregistered', () => {
      slice.unregister()
      slice.getState()
      slice.getState({ [internalStateSym]: { [sliceName]: 1 } })
      expect(console.error).toMatchSnapshot()
    })

    it('should console error when calling `setState` after unregistered', () => {
      slice.unregister()
      slice.setState(initialState, { [internalStateSym]: { [sliceName]: 1 } })
      expect(console.error).toMatchSnapshot()
    })

    it('should console error when calling `subscribe` after unregistered', () => {
      slice.unregister()
      slice.subscribe(console.log)
      expect(console.error).toMatchSnapshot()
    })

    it('should console error when calling `listen` after unregistered', () => {
      slice.unregister()
      slice.listen(console.log)
      slice.listen(slice.actions.set, console.log)
      expect(console.error).toMatchSnapshot()
    })

    it('should console error when calling action dispatcher after unregistered', () => {
      slice.unregister()
      slice.actions.set(updatedState)
      expect(console.error).toMatchSnapshot()
    })

    it('should console error when calling `reducer` after unregistered', () => {
      slice.unregister()
      slice.reducer(
        { [internalStateSym]: {} },
        slice.actions.set.create(initialState),
      )
      slice.actions.set(updatedState)
      expect(console.error).toMatchSnapshot()
    })

    it('should unregister slice reducer in parent', () => {
      slice.unregister()
      slice.actions.set(updatedState)
      expect(root.getState()).toEqual({ [internalStateSym]: {} })
    })
  })

  describe('unregistered', () => {
    it('works', () => {
      expect(slice.unregistered).toBe(false)
      slice.unregister()
      expect(slice.unregistered).toBe(true)
    })
  })
})
