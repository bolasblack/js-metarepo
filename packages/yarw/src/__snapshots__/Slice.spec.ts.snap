// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Slice > unregister > should console error when calling \`getState\` after unregistered 1`] = `
{
  "calls": [
    [
      "Calling unregistered slice "test"'s action dispatcher",
    ],
    [
      "Calling unregistered slice "test"'s action dispatcher",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Slice > unregister > should console error when calling \`listen\` after unregistered 1`] = `
{
  "calls": [
    [
      "Calling unregistered slice "test"'s action dispatcher",
    ],
    [
      "Calling unregistered slice "test"'s action dispatcher",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Slice > unregister > should console error when calling \`reducer\` after unregistered 1`] = `
{
  "calls": [
    [
      "Calling unregistered slice "test"'s action dispatcher",
    ],
    [
      "Calling unregistered slice "test"'s action dispatcher",
    ],
    [
      "Calling unregistered slice "test"'s action dispatcher",
    ],
    [
      "Calling unregistered slice "test"'s action dispatcher",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Slice > unregister > should console error when calling \`setState\` after unregistered 1`] = `
{
  "calls": [
    [
      "Calling unregistered slice "test"'s action dispatcher",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Slice > unregister > should console error when calling \`subscribe\` after unregistered 1`] = `
{
  "calls": [
    [
      "Calling unregistered slice "test"'s action dispatcher",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Slice > unregister > should console error when calling action dispatcher after unregistered 1`] = `
{
  "calls": [
    [
      "Calling unregistered slice "test"'s action dispatcher",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
